pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,230 @@
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	istiov1alpha1 "knative.dev/pkg/apis/istio/common/v1alpha1"
+	istiov1alpha3 "knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+type VirtualServiceBuilder struct {
+}
+
+func NewVirtualServiceBuilder() *VirtualServiceBuilder {
+	return &VirtualServiceBuilder{}
+}
+
+func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceStatus {
+	return &v1alpha2.VirtualServiceStatus{
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:    v1alpha2.RoutesReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  reason,
+				Message: message,
+			}},
+		},
+	}
+}
+
+func (r *VirtualServiceBuilder) CreateVirtualService(kfsvc *v1alpha2.KFService) (*istiov1alpha3.VirtualService, *v1alpha2.VirtualServiceStatus) {
+
+	httpRoutes := []istiov1alpha3.HTTPRoute{}
+
+	// TODO (rakelkar) get these from config
+	ingressGateway := "knative-ingress-gateway.knative-serving"
+	destinationServiceName := "istio-ingressgateway.istio-system.svc.cluster.local"
+
+	// destination for the default predict is required
+	predictDefaultSpec, defaultPredictReason := getPredictStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+	if predictDefaultSpec == nil {
+		return nil, createFailedStatus(defaultPredictReason, "Failed to reconcile default predictor")
+	}
+
+	// extract the virtual service hostname from the predictor hostname
+	serviceHostname := extractVirtualServiceHostname(kfsvc.Name, predictDefaultSpec.Hostname)
+	serviceURL := fmt.Sprintf("http://%s/v1/models/%s", serviceHostname, kfsvc.Name)
+
+	// add the default route
+	defaultWeight := 100 - kfsvc.Spec.CanaryTrafficPercent
+	canaryWeight := kfsvc.Spec.CanaryTrafficPercent
+	predictRouteDestinations := []istiov1alpha3.HTTPRouteDestination{
+		createHTTPRouteDestination(predictDefaultSpec.Hostname, defaultWeight, destinationServiceName),
+	}
+
+	// optionally get a destination for canary predict
+	if kfsvc.Spec.Canary != nil {
+		predictCanarySpec, canaryPredictReason := getPredictStatusConfigurationSpec(kfsvc.Spec.Canary, kfsvc.Status.Canary)
+		if predictCanarySpec != nil {
+			canaryRouteDestination := createHTTPRouteDestination(predictCanarySpec.Hostname, canaryWeight, destinationServiceName)
+			predictRouteDestinations = append(predictRouteDestinations, canaryRouteDestination)
+		} else {
+			return nil, createFailedStatus(canaryPredictReason, "Failed to reconcile canary predictor")
+		}
+	}
+
+	// prepare the predict route
+	predictRoute := istiov1alpha3.HTTPRoute{
+		Match: []istiov1alpha3.HTTPMatchRequest{
+			istiov1alpha3.HTTPMatchRequest{
+				URI: &istiov1alpha1.StringMatch{
+					Prefix: fmt.Sprintf("/v1/models/%s:predict", kfsvc.Name),
```

Comment:
any idea how can we handle the case for TensorRT and onnx since their url path won't get matched here ?

Reply:
ONNX has been fixed: https://github.com/microsoft/onnxruntime/pull/1995
(I need to update the image we reference)

TensorRT will stop working since it doesn't support the v1 format. Maybe we should mark it as experimental in the docs?

Reply:
Sounds good, we can document this and send a ticket to them to conform to our v1 protocol :).

Reply:
Updated related issue: https://github.com/kubeflow/kfserving/issues/271